core/services/gateway/connector/connector.go

```diff
@@ -0,0 +1,206 @@
+		return nil, errors.New("nil dependency")
+	}
+	if len(config.DonId) == 0 || len(config.DonId) > int(network.HandshakeDonIdLen) {
+		return nil, errors.New("invalid DON ID")
+	}
+	addressBytes, err := utils.TryParseHex(config.NodeAddress)
+	if err != nil {
+		return nil, err
+	}
+	connector := &gatewayConnector{
+		config:      config,
+		codec:       &gateway.JsonRPCCodec{},
+		clock:       clock,
+		nodeAddress: addressBytes,
+		signer:      signer,
+		handler:     handler,
+		shutdownCh:  make(chan struct{}),
+		lggr:        lggr,
+	}
+	gateways := make(map[string]*gatewayState)
+	for _, gw := range config.Gateways {
+		gw := gw
+		_, ok := gateways[gw.Id]
+		if ok {
+			return nil, fmt.Errorf("duplicate Gateway ID %s", gw.Id)
+		}
+		parsedURL, err := url.Parse(gw.URL)
+		if err != nil {
+			return nil, err
+		}
+		gateway := &gatewayState{
+			config:   &gw,
+			url:      parsedURL,
+			wsClient: network.NewWebSocketClient(config.WsClientConfig, connector, lggr),
+		}
+		gateways[gw.Id] = gateway
+	}
+	connector.gateways = gateways
+	return connector, nil
+}
+
+func (c *gatewayConnector) SendToGateway(ctx context.Context, gatewayId string, msg *gateway.Message) error {
+	data, err := c.codec.EncodeResponse(msg)
+	if err != nil {
+		return fmt.Errorf("error encoding response for gateway %s: %v", gatewayId, err)
+	}
+	gateway, ok := c.gateways[gatewayId]
+	if !ok {
+		return fmt.Errorf("invalid Gateway ID %s", gatewayId)
+	}
+	if gateway.conn == nil {
+		return fmt.Errorf("connector not started")
+	}
+	return gateway.conn.Write(ctx, websocket.BinaryMessage, data)
+}
+
+func (c *gatewayConnector) readLoop(gatewayState *gatewayState) {
+	for {
+		select {
+		case <-c.shutdownCh:
+			c.closeWait.Done()
+			return
+		case item := <-gatewayState.conn.ReadChannel():
+			msg, err := c.codec.DecodeRequest(item.Data)
+			if err != nil {
+				c.lggr.Error("parse error when reading from Gateway ", gatewayState.config.Id, err)
+				break
+			}
+			c.handler.HandleGatewayMessage(gatewayState.config.Id, msg)
+		}
+	}
+}
+
+func (c *gatewayConnector) reconnectLoop(gatewayState *gatewayState) {
+	for {
+		conn, err := gatewayState.wsClient.Connect(context.Background(), gatewayState.url)
+		if err != nil {
+			c.lggr.Error("connection error!")
+		} else {
+			closeCh := gatewayState.conn.Restart(conn)
+			<-closeCh
+			c.lggr.Info("connection closed")
+		}
+		select {
+		case <-c.shutdownCh:
+			c.closeWait.Done()
+			return
+		case <-time.After(time.Duration(c.config.ReconnectDelayMillis) * time.Millisecond):
+			c.lggr.Info("reconnecting ...")
+		}
+	}
+}
+
+func (c *gatewayConnector) Start(ctx context.Context) error {
+	return c.StartOnce("GatewayConnector", func() error {
+		c.lggr.Info("starting gateway connector")
+		for _, gatewayState := range c.gateways {
+			gatewayState.conn = network.NewWSConnectionWrapper()
+			go c.readLoop(gatewayState)
+			go c.reconnectLoop(gatewayState)
+		}
+		c.closeWait.Add(2 * len(c.gateways))
```

Comment:
For safety, `.Add` has to be done prior to spawning goroutines, otherwise they may stuck on `Done()`.